helix-view/src/document.rs

```diff
@@ -371,25 +371,54 @@ impl Document {
         encoding: Option<&'static encoding::Encoding>,
         config_loader: Option<Arc<syntax::Loader>>,
     ) -> Result<Self, Error> {
-        // Open the file if it exists, otherwise assume it is a new file (and thus empty).
-        let (rope, encoding) = if path.exists() {
+        Self::open_with_config(path, config_loader, move |_| {
+            Config::from_encoding(encoding)
+        })
+    }
+
+    // TODO: async fn?
+    /// Create a new document from `path`,
+    /// using document settings from a [Config].
+    pub fn open_with_config(
+        path: &Path,
+        lang_config_loader: Option<Arc<syntax::Loader>>,
+        doc_config_fn: impl FnOnce(&Path) -> Config,
+    ) -> Result<Self, Error> {
+        let (rope, config) = if path.exists() {
             let mut file =
                 std::fs::File::open(path).context(format!("unable to open {:?}", path))?;
-            from_reader(&mut file, encoding)?
+            let mut config = doc_config_fn(path);
```

Comment:
The reason I call `doc_config_fn` there and not before the `if path.exists()` line is to avoid parsing the EditorConfig files if opening an existing file fails.

That said, I did have a thought that the logic for opening a `Document` could look something like this, which should also be tidier:
1. Attempt to open with `File::open`. Don't check if the path exists first.
2. If opening fails, continue only if the error is not [`NotFound`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.NotFound)
3. Determine the encoding. This may involve parsing the EditorConfig files at this point.
4. Construct the rope using the `File` (if open) and the encoding.
5. Construct the `Config` now, falling back to auto-detection from the rope's contents.
6. Finally construct and return the `Document`.

Reply:
Wouldn't it be better to move this out of `open` and put it beside open instead of changing the open function directly?

Reply:
What I meant is something like the below in `Editor::open`
```rust
            ...
            let file = File::open(...)?;
            let config = self.editorconfig(file);
            let doc = Document::open(&file, None, syn_loader, config)?;
```
Like move the structure around a bit, rather than having both stuck at the same place.

So now instead of Document open taking in the path, it takes a file, or it could take an enum or something else.